staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -569,6 +605,50 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
 	return nil
 }
 
+// IsStarted reports whether the informer has already been started
+func (s *sharedIndexInformer) IsStarted() bool {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+	return s.started
+}
+
+// IsStopped reports whether the informer has already been stopped
+func (s *sharedIndexInformer) IsStopped() bool {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+	return s.stopped
+}
+
+// EventHandlerCount reports whether the informer still has registered
+// event handlers
+func (s *sharedIndexInformer) EventHandlerCount() int {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+	return len(s.processor.listeners)
+}
+
+// RemoveEventHandlerByHandle tries to remove a formerly added event handler by its
+// handle returned for its registration.
+// If a handler has been added multiple times, only the registration for the
+// given handle will be removed.
+func (s *sharedIndexInformer) RemoveEventHandlerByHandle(handle *ResourceEventHandlerHandle) error {
+	s.startedLock.Lock()
+	defer s.startedLock.Unlock()
+
+	if s.stopped {
+		return fmt.Errorf("handler %v is not removed from shared informer because it has stopped already", handle.ResourceEventHandler)
+	}
+
+	// in order to safely remove, we have to
+	// 1. stop sending add/update/delete notifications
+	// 2. remove and stop listener
+	// 3. unblock
+	s.blockDeltas.Lock()
+	defer s.blockDeltas.Unlock()
+	s.processor.removeListenerFor(handle)
```

Comment:
Good idea, because now single registrations and not handlers are removed the complete handling of removing multiple listeners get obsolete. I'll change this later. Now I've to leave  for two hours.

Please give me an advise, which handle model (see above) you prefer. I've already implemented the second solution, but not yet pushed it.

Reply:
> make the field private and implement the handler interface for the handle (that's what I've done with the next commit)

I don't think the handle we return should implement the handler interface.

> add another invisible indirection layer by internally registering a wrapped handler and return a handle referring to the wrapped handler to be able to remove it again

something along those lines sounds good... I think the listener we already created internally can serve as the wrapped handler, right?

Reply:
Yepp, directly using the listener as proposed by @liggitt instead of a wrapper just implies the second behaviour. Now the handle is just an anonymous handle for the outside world and directly refers to the created listener already wrapping the handler.